### PR TITLE
Encapsulate promises in async functions

### PR DESCRIPTION
**Pull request by ychoquet**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # This PR is for investigation

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

__Deploy URL__: https://ychoquet.github.io/GeoView/type-of-layers.html

# Checklist:

- [x] I have build __(rush build)__ and deploy __(rush host)__ my PR
- [ ] I have connected the issues(s) to this PR
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have created new issue(s) related to the outcome of this PR is needed
-  ~~I have made corresponding changes to the documentation~~
-  ~~I have added tests that prove my fix is effective or that my feature works~~
-  ~~New and existing unit tests pass locally with my changes~~

- - -
This change is [Reviewable](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/1417)

### CODE REVIEW

**Review by Alex-NRCan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/1417)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @ychoquet)

___
*[`packages/geoview-core/src/geo/layer/geoview-layers/abstract-geoview-layers.ts` line 604 at r2](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/1417#-NhDSR5XDkfL50Mbt-Pn:-NhDSR5XDkfL50Mbt-Po:b-85o38j) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/917bb8da47a61b879e50d1f3b403f2832d750460/packages/geoview-core/src/geo/layer/geoview-layers/abstract-geoview-layers.ts#L604)):*
> ```XML
>    * @returns {Promise<void>} A promise that the vector layer configuration has its metadata processed.
>    */
>   protected async processLayerMetadata(layerEntryConfig: TypeLayerEntryConfig): Promise<void> {
> ```

If there's no 'await' needed for a particular function implementation, the function shouldn't use an async signature and just return void, not a promise.
Same comment for couple other functions like e.g. getAllFeatureInfo, getFeatureInfoAtPixel below.

**Review by Alex-NRCan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/1417)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @ychoquet)

___
*[`packages/geoview-core/src/geo/layer/geoview-layers/abstract-geoview-layers.ts` line 561 at r2](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/1417#-NhDwb0M4HT1ARj_Auzc:-NhDwb0N1ghw6j53taYw:bd03rhs) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/917bb8da47a61b879e50d1f3b403f2832d750460/packages/geoview-core/src/geo/layer/geoview-layers/abstract-geoview-layers.ts#L561)):*
> ```XML
>    */
>   protected async processListOfLayerEntryMetadata(
>     listOfLayerEntryConfig: TypeListOfLayerEntryConfig = this.listOfLayerEntryConfig
> ```

I've read the code and in this case, using an *optional parameter* pointing to `this.listOfLayerEntryConfig` seems to be 'thread safe'. Therefore, this comment has a no-action with it. So, please take this comment more as an open discussion than anything.

I just want to take the opportunity to describe a concern of mine with parallel code and usage of optional parameters that references a `this.someProperty`.

Consider this code example

This will print a console showing worker B working on a changing value, mid-execution. There can be good cases for this, but it's a bit risky, in general.



_Code snippet (i):_
```XML
class TEST {
  value?: string;

  sleepAsync = (ms: number) => {
    return new Promise((r) => {
      setTimeout(r, ms);
    });
  };

  doWorkOnValue = (key: string, value: string = this.value!): void => {
    console.log(`=> ${key} is working using the value: ${value}`);
  };

  simulateSomeWorkHappeningChangingTheValueAsync = async (delay: number) => {
    await this.sleepAsync(delay);
    this.value = 'salut';
  };

  doWorkAsync = async (key: string, delay: number): Promise<void> => {
    console.log(`===> ${key} started working with the value: ${this.value}`);

    // Call the function which uses this.value as default param
    this.doWorkOnValue(key);

    // Pretend to work for some time (delay)
    await this.sleepAsync(delay);

    // Call the function which uses this.value as default param
    this.doWorkOnValue(key);
    console.log(`===> ${key} finished working with the value: ${this.value}`);
  };

  tryMe = () => {
    // We want to use the value bonjour for this parallel processing
    this.value = 'bonjour';

    // Pretend a user interaction which will change the value of 'this.value' in 2000ms
    this.simulateSomeWorkHappeningChangingTheValueAsync(2000);

    // Start tasks using various ms durations, all expecting stuff to happen with value = bonjour
    this.doWorkAsync('A', 1000);
    this.doWorkAsync('B', 3000);
    this.doWorkAsync('C', 0);

    // A and C will complete their work before 'this.value' got tampered with
    // C will finish its work first, because it's fast
    // B, because they take longer than 2000 to run, will start with the original 'this.value'
    // and see it change in the middle of its execution
  };
}

export function RUN_TEST() {
  const test = new TEST();
  test.tryMe();
}
```



_Code snippet (ii):_
```XML
===> A started working with the value: bonjour
=> A is working using the value: bonjour
===> B started working with the value: bonjour
=> B is working using the value: bonjour               <------here
===> C started working with the value: bonjour
=> C is working using the value: bonjour
=> C is working using the value: bonjour
===> C finished working with the value: bonjour
=> A is working using the value: bonjour
===> A finished working with the value: bonjour
=> B is working using the value: salut                 <------here
===> B finished working with the value: salut
```

**Review by Alex-NRCan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/1417)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @ychoquet)

___
*[`packages/geoview-core/src/geo/layer/geoview-layers/abstract-geoview-layers.ts` line 561 at r2](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/1417#-NhDwb0M4HT1ARj_Auzc:-NhE2b-p6unfz_XU4HXe:b-w2lytw) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/917bb8da47a61b879e50d1f3b403f2832d750460/packages/geoview-core/src/geo/layer/geoview-layers/abstract-geoview-layers.ts#L561)):*
<details><summary><i>Previously, Alex-NRCan (Alex) wrote…</i></summary><blockquote>

I've read the code and in this case, using an *optional parameter* pointing to `this.listOfLayerEntryConfig` seems to be 'thread safe'. Therefore, this comment has a no-action with it. So, please take this comment more as an open discussion than anything.

I just want to take the opportunity to describe a concern of mine with parallel code and usage of optional parameters that references a `this.someProperty`.

Consider this code example

This will print a console showing worker B working on a changing value, mid-execution. There can be good cases for this, but it's a bit risky, in general.
</blockquote></details>

I admit, removing the optional parameter doesn't fix the B worker from working on 'salut', but it's a bit clearer to read the code and figure it out.
Basically, async is great, but extra caution needs to be given when using 'this' anywhere.

**Review by ychoquet — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/1417)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @Alex-NRCan)

___
*[`packages/geoview-core/src/geo/layer/geoview-layers/abstract-geoview-layers.ts` line 561 at r2](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/1417#-NhDwb0M4HT1ARj_Auzc:-NhRq5WP55tVN1SIWkU9:bcz8nl5) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/06c34602b150dbeac0194699baf1e08422039a10/packages/geoview-core/src/geo/layer/geoview-layers/abstract-geoview-layers.ts#L561)):*
<details><summary><i>Previously, Alex-NRCan (Alex) wrote…</i></summary><blockquote>

Removing the optional parameter doesn't fix the B worker from working on 'salut', but it's a bit clearer to read the code and figure it out.
Basically, async is great, but extra caution needs to be given when using 'this' anywhere.
</blockquote></details>

I made the correction to be safer..
___
*[`packages/geoview-core/src/geo/layer/geoview-layers/abstract-geoview-layers.ts` line 604 at r2](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/1417#-NhDSR5XDkfL50Mbt-Pn:-NhSkfVN5mEzgyj54eMo:b-27ii7d) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/06c34602b150dbeac0194699baf1e08422039a10/packages/geoview-core/src/geo/layer/geoview-layers/abstract-geoview-layers.ts#L604)):*
<details><summary><i>Previously, Alex-NRCan (Alex) wrote…</i></summary><blockquote>

If there's no 'await' needed for a particular function implementation, the function shouldn't use an async signature and just return void, not a promise.
Same comment for couple other functions like e.g. getAllFeatureInfo, getFeatureInfoAtPixel below.
</blockquote></details>

Here, async is used because some children cannot execute this method synchronously. We want to have a homogeneous definition of the method in the inheritance tree.

**Review by Alex-NRCan — COMMENTED** [review]
I see the suffix 'Async' suffix isn't being used, is it on purpose?
I believe we should either use the suffix or not, throughout GeoView core.

My opinion is that we should be using the suffix for clarity, but if we're not using it, I'd remove it from other places - so that we're at least coherent when developers work in the project. Instead of having to check the function signature each time.

> Reviewed 1 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/1417#-:-NhSxw_01dF3-5ISvoBa:ba7q7xo)* status: 1 of 2 files reviewed, 1 unresolved discussion (waiting on @ychoquet)

___
*[`packages/geoview-core/src/geo/layer/geoview-layers/abstract-geoview-layers.ts` line 604 at r2](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/1417#-NhDSR5XDkfL50Mbt-Pn:-NhSxNoq9EaPdDKW_F4v:b-r3g280) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/06c34602b150dbeac0194699baf1e08422039a10/packages/geoview-core/src/geo/layer/geoview-layers/abstract-geoview-layers.ts#L604)):*
<details><summary><i>Previously, ychoquet (Yves Choquette) wrote…</i></summary><blockquote>

Here, async is used because some children cannot execute this method synchronously. We want to have a homogeneous definition of the method in the inheritance tree.
</blockquote></details>

Not sure what you mean, let's talk about it a bit please

**Review by Alex-NRCan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/1417)* status: 1 of 2 files reviewed, 1 unresolved discussion (waiting on @ychoquet)

___
*[`packages/geoview-core/src/geo/layer/geoview-layers/abstract-geoview-layers.ts` line 604 at r2](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/1417#-NhDSR5XDkfL50Mbt-Pn:-NhTMK-RCUMZE8oGZFx4:buc6oeo) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/06c34602b150dbeac0194699baf1e08422039a10/packages/geoview-core/src/geo/layer/geoview-layers/abstract-geoview-layers.ts#L604)):*
<details><summary><i>Previously, Alex-NRCan (Alex) wrote…</i></summary><blockquote>

Not sure what you mean, let's talk about it a bit please
</blockquote></details>

After discussion, return `Promise<void>` being returned is important, because that's the signature of the method as it should be overridden. We can't turn it into return void - like my first comment was saying as some child class will be returning actual Promises.

My other suggestion was to remove 'async' keyword as it's superfluous in this case - as there's no `await` in the implementation of the function for this particular class. However, doing so would force implementation to "actually" return a Promise (it'd require a bit more lines of code) to actually return a Promise instead of just `void`. It is allowed by JavaScript to have an `async`keyword for a function that just returns `void` (I guess it creates a Promise somewhere without saying it).

That comment is debatable. I can agree to either leave it as is:

or add 1 line that explicitly returns a Promise (and remove the 'async' keyword).



_Code snippet (i):_
```XML
   if (!layerEntryConfig.source) layerEntryConfig.source = {};
   if (!layerEntryConfig.source.featureInfo) layerEntryConfig.source.featureInfo = { queryable: true };
```



_Code snippet (ii):_
```XML
   if (!layerEntryConfig.source) layerEntryConfig.source = {};
   if (!layerEntryConfig.source.featureInfo) layerEntryConfig.source.featureInfo = { queryable: true };
   return Promise.resolve();
```

**Review by ychoquet — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/1417)* status: 1 of 2 files reviewed, 1 unresolved discussion (waiting on @Alex-NRCan)

___
*[`packages/geoview-core/src/geo/layer/geoview-layers/abstract-geoview-layers.ts` line 604 at r2](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/1417#-NhDSR5XDkfL50Mbt-Pn:-NhTosal5wzvttVhH81a:b-ei51y7) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/06c34602b150dbeac0194699baf1e08422039a10/packages/geoview-core/src/geo/layer/geoview-layers/abstract-geoview-layers.ts#L604)):*
<details><summary><i>Previously, Alex-NRCan (Alex) wrote…</i></summary><blockquote>

After discussion, return `Promise<void>` being returned is important, because that's the signature of the method as it should be overridden. We can't turn it into return void - like my first comment was saying as some child class will be returning actual Promises.

My other suggestion was to remove 'async' keyword as it's superfluous in this case - as there's no `await` in the implementation of the function for this particular class. However, doing so would force implementation to "actually" return a Promise (it'd require a bit more lines of code) to actually return a Promise instead of just `void`. It is allowed by JavaScript to have an `async`keyword for a function that just returns `void` (I guess it creates a Promise somewhere without saying it).

That comment is debatable. I can agree to either leave it as is:

or add 1 line that explicitly returns a Promise (and remove the 'async' keyword).
</blockquote></details>

As far as I'm concerned, as long as the code does what we want it to do, I have no preference.

**Review by jolevesq — COMMENTED** [review]
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/1417)* status: 1 of 2 files reviewed, 1 unresolved discussion (waiting on @Alex-NRCan)

___
*[`packages/geoview-core/src/geo/layer/geoview-layers/abstract-geoview-layers.ts` line 604 at r2](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/1417#-NhDSR5XDkfL50Mbt-Pn:-NhWoJ24BO_2IG4KVdVl:b-tqcybi) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/06c34602b150dbeac0194699baf1e08422039a10/packages/geoview-core/src/geo/layer/geoview-layers/abstract-geoview-layers.ts#L604)):*
<details><summary><i>Previously, ychoquet (Yves Choquette) wrote…</i></summary><blockquote>

As far as I'm concerned, as long as the code does what we want it to do, I have no preference.
</blockquote></details>

I think we should explicitly return Promise.resolve() as it follows the function signature and it is cleaner.

**Review by jolevesq — APPROVED** [review]
> Reviewed 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/1417)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @ychoquet)
